consensus/core/src/commit_observer.rs

```diff
@@ -0,0 +1,471 @@
+        let _observer = CommitObserver::new(
+            context.clone(),
+            linearizer,
+            sender,
+            expected_last_processed_index as CommitIndex,
+            mem_store.clone(),
+        );
+
+        // Check commits sent over consensus output channel is accurate starting
+        // from last processed index of 2 and finishing at last sent index of 3.
+        processed_subdag_index = expected_last_processed_index;
+        while let Ok(subdag) = receiver.try_recv() {
+            tracing::info!("Processed {subdag} on resubmission");
+            assert_eq!(subdag, commits[processed_subdag_index]);
+            processed_subdag_index = subdag.commit_index as usize;
+            if processed_subdag_index == expected_last_sent_index {
+                break;
+            }
+        }
+
+        verify_channel_empty(&mut receiver);
+    }
+
+    #[test]
+    fn test_send_no_missing_commits() {
+        telemetry_subscribers::init_for_testing();
+        let num_authorities = 4;
+        let context = Arc::new(Context::new_for_test(num_authorities).0);
+        let mem_store = Arc::new(MemStore::new());
+        let dag_state = Arc::new(RwLock::new(DagState::new(
+            context.clone(),
+            mem_store.clone(),
+        )));
+        let linearizer = Linearizer::new(dag_state.clone());
+        let leader_schedule = LeaderSchedule::new(context.clone());
+        let last_processed_index = 0;
+        #[allow(clippy::disallowed_methods)] // allow unbounded_channel()
+        let (sender, mut receiver) = unbounded_channel();
+
+        let mut observer = CommitObserver::new(
+            context.clone(),
+            linearizer.clone(),
+            sender.clone(),
+            last_processed_index,
+            mem_store.clone(),
+        );
+
+        // Populate fully connected test blocks for round 0 ~ 10, authorities 0 ~ 3.
+        let num_rounds = 10;
+        build_dag(context.clone(), dag_state.clone(), None, num_rounds);
+        let leaders = get_all_leader_blocks(
+            dag_state.clone(),
+            leader_schedule,
+            num_rounds,
+            DEFAULT_WAVE_LENGTH,
+            false,
+            1,
+        );
+
+        // Commit all of the leaders and "receive" the subdags as the consumer of
+        // the consensus output channel.
+        let expected_last_processed_index = 3;
+        let commits = observer.handle_commit(leaders.clone());
+
+        // Check commits sent over consensus output channel is accurate
+        let mut processed_subdag_index = 0;
+        while let Ok(subdag) = receiver.try_recv() {
+            tracing::info!("Processed {subdag}");
+            assert_eq!(subdag, commits[processed_subdag_index]);
+            processed_subdag_index = subdag.commit_index as usize;
+            if processed_subdag_index == expected_last_processed_index {
+                break;
+            }
+        }
```

Comment:
I believe might need to assert after the loop that the `processed_subdag_index` has been updated to the latest expected value as there is the case where the channel is from the beginning empty. In this scenario the `receiver.try_recv` will just return an `Err` for `Empty` channel directly and the loop will never process. I see that we do verify the last stored commit , but I think it would be good to make sure that the loop did run indeed